traitsui/examples/demo/Standard_Editors/ListEditor_demo.py

```diff
@@ -21,13 +21,13 @@ class ListEditorDemo(HasTraits):
 
     # Items are used to define display, one per editor style:
     list_group = Group(
-        Item('play_list', style='simple', label='Simple', height=75),
+        Item('play_list', style='simple', label='Simple', height=75, id='simple'),
         Item('_'),
-        Item('play_list', style='custom', label='Custom'),
+        Item('play_list', style='custom', label='Custom', id='custom'),
         Item('_'),
-        Item('play_list', style='text', label='Text'),
+        Item('play_list', style='text', label='Text', id='text'),
         Item('_'),
-        Item('play_list', style='readonly', label='ReadOnly')
+        Item('play_list', style='readonly', label='ReadOnly', id='readonly')
```

Comment:
One thing I found very strange/misleading is that these styles of editor dont really exist.  All of these are exactly the same and in fact, in the demo a user can modify the 'readonly' one just like the rest.  The code for getting the list editor is simply:
```
def _get_custom_editor_class(self):
        if self.use_notebook:
            return toolkit_object("list_editor:NotebookEditor")
        return toolkit_object("list_editor:CustomEditor")
```
This is not directly relevant to this PR, but it seemed off to me

Reply:
Note they are defined:
https://github.com/enthought/traitsui/blob/5a030766dd1168ddeb14f2d481a08e484e65f763/traitsui/qt4/list_editor.py#L412-L423
https://github.com/enthought/traitsui/blob/5a030766dd1168ddeb14f2d481a08e484e65f763/traitsui/wx/list_editor.py#L473-L484
but based on the above code it seems they aren't ever going to be used?

Reply:
You are right to find this strange.

The design of the EditorFactory is that there are four allowed values for "style": "simple", "custom", "text", "readonly". (defined [here](https://github.com/enthought/traitsui/blob/6f25745e3a9035b17e5512998dbc76b91be53f2d/traitsui/ui_traits.py#L56-L57)).
And each style should map to one editor class. The default implementation is like this:
https://github.com/enthought/traitsui/blob/460742421bcdddfec124a35287843e8a0eae0b49/traitsui/editor_factory.py#L222-L283

Trouble is, sometimes we need more than four editor classes for an ever growing number of options, but the allowed values in "style" is fixed. So the editor factory ends up having to branch out from one of these four styles (usually "custom", for its name is the least specific).

e.g. RangeEditor has this `mode` option, and to select an editor class, you need to combine "style" and "mode":
https://github.com/enthought/traitsui/blob/460742421bcdddfec124a35287843e8a0eae0b49/traitsui/editors/range_editor.py#L216-L271

Reply:
That makes sense, thank you!

So, because the ToolkitEditorFactory in https://github.com/enthought/traitsui/blob/5a030766dd1168ddeb14f2d481a08e484e65f763/traitsui/editors/list_editor.py#L170-L173 
subclasses EditorFactory, it has the `_get_simple_editor_class`, `_get_custom_editor_class`, `_get_text_editor_class`, and `_get_readonly_editor_class` methods.  The only one overridden is `_get_custom_editor_class` where there is another check added to see if someone wants to `use_notebook` and they can't add another style because only the original 4 are allowed.  My initial source of confusion was that all 4 styles appear exactly the same in the demo.  The one that stood out as seeming wrong was the readonly style, as if you run the demo, you can freely modify the text, and see the trait is updated as the other styles all immediately reflect any changes you make.  This didn't seem very 'readonly' and I initially thought for that it wasn't actually the ReadonlyEditor being used.  After looking more closely I realize that this actually intentional as the list being readonly only sets the mutable trait to false.  Thus the readonly style does not provide the additional buttons to 'Add Before' or 'Add After' etc, but the things in the list can still be modified.